server/src/main/java/org/elasticsearch/transport/RemoteClusterCredentialsManager.java

```diff
@@ -13,32 +13,46 @@
 import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.settings.SecureString;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.set.Sets;
 import org.elasticsearch.core.Nullable;
 
+import java.util.Collections;
 import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
 
 import static org.elasticsearch.transport.RemoteClusterService.REMOTE_CLUSTER_CREDENTIALS;
 
 public class RemoteClusterCredentialsManager {
-
     private static final Logger logger = LogManager.getLogger(RemoteClusterCredentialsManager.class);
 
-    private volatile Map<String, SecureString> clusterCredentials;
+    private volatile Map<String, SecureString> clusterCredentials = Collections.emptyMap();
 
     public RemoteClusterCredentialsManager(Settings settings) {
         updateClusterCredentials(settings);
     }
 
-    public final void updateClusterCredentials(Settings settings) {
-        clusterCredentials = REMOTE_CLUSTER_CREDENTIALS.getAsMap(settings);
-        logger.debug(
-            () -> Strings.format(
-                "Updated remote cluster credentials for clusters: [%s]",
-                Strings.collectionToCommaDelimitedString(clusterCredentials.keySet())
-            )
-        );
+    public final synchronized UpdateRemoteClusterCredentialsResult updateClusterCredentials(Settings settings) {
```

Comment:
This is only called from a single-threaded context. Might consider keeping it synchronized  for future proofing, but it adds some overhead to acquire an extra lock.

Reply:
Yes great point. I meant to take a closer look at being less heavy handed with synchronization but didn't get to it. Thanks for prompting me to think about this again. I agree that with the current structure, we _can_ remove synchronization here. I'll think on it and tweak some things 👍

Reply:
Discussed this on Slack also:

* I prefer to keep this one `synchronized` for future proofing
* I removed sorting the cluster aliases as that adds unnecessary overhead without utility -- originally, I used sorting with the notion to prevent deadlocks in `RemoteClusterService::updateRemoteClusterCredentials` (since we iterate over aliases and call a synchronized method for each, we have a potential deadlock). However `updateRemoteClusterCredentials` is itself syncronized at the top level, so we're safe.

Thanks for raising and discussing this 👍